File: Exercises/rb101-rb109_small_problems/easy_8/all_substrings.rb
# frozen_string_literal: true

def leading_substrings(string)
  arr = []
  0.upto(string.size - 1) do |num|
    arr << string[0..num]
  end
  arr
end

def substrings(string)
  arr = []
  count = 0
  loop do
    arr.concat(leading_substrings(string[count..(string.size - 1)]))
    break if count == string.size - 1

    count += 1
  end
  arr
end

puts substrings('abcde') == [
  'a', 'ab', 'abc', 'abcd', 'abcde',
  'b', 'bc', 'bcd', 'bcde',
  'c', 'cd', 'cde',
  'd', 'de',
  'e'
]

File: ls215/exercises/small_problems/list_processing/all_substrings.js
function leadingSubstrings(string) {
  const chars = string.split('');

  return chars.map((char, index) => {
    return chars.slice(0, index + 1).join('');
  });
}

function substrings(string) {
  const chars = string.split('');

  return chars.flatMap((char, index) => {
    return leadingSubstrings(chars.slice(index).join(''));
  });
}

console.log(substrings('abcde'));

// returns
[ "a", "ab", "abc", "abcd", "abcde",
  "b", "bc", "bcd", "bcde",
  "c", "cd", "cde",
  "d", "de",
  "e" ]
